codalab/migration.py

```diff
@@ -0,0 +1,216 @@
+        return self.bundle_manager._model.get_bundle(bundle_uuid)
+
+    def get_bundle_info(self, bundle_uuid, bundle_location):
+        target = BundleTarget(bundle_uuid, subpath='')
+        logging.info(f"[migration] {target}")
+        try:
+            info = download_util.get_target_info(bundle_location, target, depth=0)
+            logging.info(f"[migration] {info}")
+        except Exception as e:
+            logging.info(f"[migration] Error: {str(e)}")
+            raise e
+
+        return info
+
+    def upload_to_Azure(self, bundle_uuid, bundle_location, is_dir=False):
+        # generate target bundle path
+        file_name = "contents.tar.gz" if is_dir else "contents.gz"
+        target_location = f"{self.target_store_url}/{bundle_uuid}/{file_name}"
+
+        if FileSystems.exists(target_location):
+            path_util.remove(target_location)
+
+        uploader = BlobStorageUploader(
+            bundle_model=self.bundle_manager._model,
+            bundle_store=self.bundle_manager._bundle_store,
+            destination_bundle_store=None,
+            json_api_client=None,
+        )
+
+        if is_dir:
+            source_fileobj = zip_util.tar_gzip_directory(bundle_location)
+            source_ext = ".tar.gz"
+            unpack = True
+        else:
+            # If it's a file, change it into GzipStream
+            source_fileobj = open(bundle_location, 'rb')
+            source_ext = ''
+            unpack = False
+
+        # Upload file content and generate index file
+        uploader.write_fileobj(source_ext, source_fileobj, target_location, unpack_archive=unpack)
+
+        assert FileSystems.exists(target_location)
+
+    def modify_bundle_data(self, bundle, bundle_uuid, is_dir):
+        """
+        Change the bundle location in the database
+        ATTENTION: this function will modify codalab
+        """
+
+        # add bundle location
+        self.bundle_manager._model.add_bundle_location(bundle_uuid, self.target_store_uuid)
+
+        new_location = self.get_bundle_location(bundle_uuid)
+        assert new_location.startswith(StorageURLScheme.AZURE_BLOB_STORAGE.value)
+
+        # Update metadata
+        metadata = self.bundle_manager._model.get_bundle_metadata(
+            uuids=[bundle_uuid], metadata_key={'store'}
+        )
+        assert metadata.get('store', None) is None
+
+        # storage_type is a legacy field, will still update this field because there is no side effect
+        self.bundle_manager._model.update_bundle(
+            bundle,
+            {
+                'storage_type': self.target_store_type.value,
+                'is_dir': is_dir,
+                'metadata': {'store': self.target_store_name},
+            },
+        )
+
+    def delete_origin_bundle(self, bundle_uuid, bundle_location):
+        # Delete data from orginal bundle store
+        if os.path.exists(bundle_location):
+            path_util.remove(bundle_location)
+
```

Comment:
If you want to add the disk quota decrement here, you can simply do:
```
removed = path_util.remove(bundle_location)
if removed:
  data_size = self.bundle_manager._model.get_bundle_metadata(bundle_uuid, 'data_size')[bundle_uuid]
   self.bundle_manager._model.increment_user_disk_used(
                    user_id, -int(data_size)
                )
```
bundle_data_sizes = [bundle_uuid]

Reply:
Wow thanks Andrew! Since the old bundle metadata is already modified here, so we can not use this size. But thanks a lot for reminding me function `increment_user_disk_used`!